### ISSUE TITLE
Cameras not recording

### ISSUE THREAD

**Issue by rcho5840** [motioneye-project/motioneye#3023]
I have tried to use motion detection recording and although I see the red outline, there are no videos that are saved. Now, keep in mind that I do not have a flash drive in. My motioneye is setup on a raspberry pi with a microsd card. I also have a flashdrive with a custom path set to the flash drive storage
![image](https://github.com/motioneye-project/motioneye/assets/134464812/316d731c-c6bf-44cc-be58-06036578d322)
![image](https://github.com/motioneye-project/motioneye/assets/134464812/251aa775-96ee-4722-bf7f-4614c1a10159)

The second issue I noticed was that even with the continuous recording, there is no red outline. I believe this means that it's just not recording in general.

![Screenshot 2024-07-01 110953](https://github.com/motioneye-project/motioneye/assets/134464812/c891ac1c-9a1c-45dd-b714-1bffdcd83462)

**Comment by Martypose** [comment]
change movies to this

![config](https://github.com/motioneye-project/motioneye/assets/46140279/051e04b4-1437-49cf-8b46-d58b9c25a5fc)

change the codec, that was my problem, default didnt work

**Comment by rcho5840** [comment]
@Martypose yup one of the issues was the codec. It seems for my camera, it only wants to use mpeg however, so there's that.

**Comment by zagrim** [comment]
Recording mode "Continuous recording" should mean just what it says, and it actually does the same as adding `emulate_motion on` in Extra Motion Options. And it should show the red outline.

Could you @rcho5840 please post the logs (`sudo journalctl -u motion`) so we can see if there are some underlying issues with your config?

Please also clarify this part:
> Now, keep in mind that I do not have a flash drive in. My motioneye is setup on a raspberry pi with a microsd card. I also have a flashdrive with a custom path set to the flash drive storage

Do you mean that you have not connected the flash drive, but you still have the flash drive path set in "File storage" > "Root directory"? That would obviously not work, but maybe I read your words wrong :smiley:

**Comment by zagrim** [comment]
I also noticed that you'd opened #3025 and I wonder if that was resolved and if these two issues are actually related?

**Comment by Matthew1471** [comment]
I came here to raise that OMX doesn't work too.. I got errors from motion saying it couldn't set the codec but ffmpeg does seem to have OMX compiled in.

**Comment by MichaIng** [comment]
OMX depends on legacy RPi-only APIs which are not available anymore since Bookworm. And yes, `h264_v4l2m2m` the hardware accelerated successor if supported by the board, else plain `h264` (or a different codec).

**Comment by ToeKneeFan** [comment]
I know this is old/closed, but I thought I would share in case it helps anyone else. (And my problem was the same as this issue.)

At least for me, the standard ffmpeg was not working properly with `h264_v4l2m2m` out of the box and was issuing errors. Switching to H.264 fixes this, but that forces software encoding, resulting in poor performance. However, manually compiling jc-kynesim's [ffmpeg fork](https://github.com/jc-kynesim/rpi-ffmpeg/) with patches/revisions for Raspberry PI and using that instead worked.

This is actually not unlike OMX back in the day (when that was previously supported in the legacy APIs).  Users needed to patch and manually compile it ([1](https://github.com/motioneye-project/motioneye/issues/930#issuecomment-522848452)). For that reason, a nonstandard patched version of ffmpeg was incorporated into motioneyeos.

**Comment by MichaIng** [comment]
Did the FFmpeg build from RPi OS not work (as good)? Weird that a source that has not seen any commit since 3 years does 🤔.

But yeah, FFmpeg != FFmpeg. There is also this V4L2 Request API support outstanding in mainline FFmpeg, needed for hardware acceleration with a bunch of SoCs: https://code.ffmpeg.org/FFmpeg/FFmpeg/pulls/20847

**Comment by ToeKneeFan** [comment]
@MichaIng  I use Diet Pi (that you maintain, I believe, thank you 🙏🏻), but if I recall correctly, it pulls the Raspbian / Raspberry Pi OS distribution's ffmpeg, so yes, it does not work (at least on my hardware, Pi Zero W and Pi Model 3A+). Apologies, I should have clarified that I used the test/7.1.2/main branch of rpi-ffmpeg, which is more recently updated. It is one of the branch's suggested by that repo's maintainer.

Using the more mainline ffmpeg, the camera view is visible in the UI, but it fails to write to file on the default setting of "H264/V4L2M2M". (Same issue as reported by OP in this issue.) Like OP, it recovers when switched to the "H264" setting, though of course encoding performance (namely fps) is sacrificed. Even though `h264_v4l2m2m` was listed as a supported encoder by `ffmpeg -encoders`, it would fail and emit errors similar to [this comment](https://github.com/motioneye-project/motioneye/issues/3023#issuecomment-2518604852). 

I ended up switching to manually compiled rpi-ffmpeg/test/7.1.2/main, which fixed the issue and dramatically improved performance. (from ~1 fps to ~30 fps)

**Comment by MichaIng** [comment]
Yeah, DietPi on RPi uses the RPi repository, which ships an adjusted FFmpeg build. Since you use DietPi, please assure the following:
- In `dietpi-config` display options > resolution select `vc4-kms-v3d`.
- Do NOT enable camera support there. It is the legacy one, which we really need to remove ASAP, as we need to remove the old resolution menu, make KMS a dedicated option, and resolution changes via `dietpi-display` only.
- Instead enable RPi Codecs

And run this:
```sh
rm /etc/modprobe.d/dietpi-disable_rpi_camera.conf
G_CONFIG_INJECT 'camera_auto_detect=' 'camera_auto_detect=1' /boot/firmware/config.txt
```

**Comment by ToeKneeFan** [comment]
@MichaIng 
> Do NOT enable camera support there.

Do you mean have Display Options > RPi Camera turned [Off]?

If so, this results in
```
[1:ml1:Camera1] [ALR] [VID] v4l2_device_open: Failed to open video device /dev/video0: No such file or directory
```

Making the other changes you suggested results in this:
```
motion[1065]: [0:motion] [NTC] [ALL] motion_startup: Motion 4.7.0 Started
motion[1065]: [0:motion] [NTC] [ALL] motion_startup: Using default log type (ALL)
motion[1065]: [0:motion] [NTC] [ALL] motion_startup: Using log type (ALL) log level (WRN)
motion[1065]: [1:ml1:Camera1] [ERR] [VID] v4l2_device_capability: Not a V4L2 device?
motion[1065]: [1:ml1:Camera1] [ERR] [VID] vid_start: V4L2 device failed to open
motion[1065]: [1:ml1:Camera1] [WRN] [ALL] motion_init: Could not fetch initial image from camera
motion[1065]: [1:ml1:Camera1] [WRN] [ALL] motion_init: Motion continues using width and height from config file(s)
motion[1065]: [1:ml1:Camera1] [WRN] [ALL] mlp_retry: Retrying until successful connection with camera
motion[1065]: [1:ml1:Camera1] [ERR] [VID] v4l2_device_capability: Not a V4L2 device?
motion[1065]: [1:ml1:Camera1] [ERR] [VID] vid_start: V4L2 device failed to open
```

**Comment by ToeKneeFan** [comment]
> I know this is old/closed, but I thought I would share in case it helps anyone else. (And my problem was the same as this issue.)
> 
> At least for me, the standard ffmpeg was not working properly with `h264_v4l2m2m` out of the box and was issuing errors. Switching to H.264 fixes this, but that forces software encoding, resulting in poor performance. However, manually compiling jc-kynesim's [ffmpeg fork](https://github.com/jc-kynesim/rpi-ffmpeg/) with patches/revisions for Raspberry PI and using that instead worked.
> 
> This is actually not unlike OMX back in the day (when that was previously supported in the legacy APIs). Users needed to patch and manually compile it ([1](https://github.com/motioneye-project/motioneye/issues/930#issuecomment-522848452)). For that reason, a nonstandard patched version of ffmpeg was incorporated into motioneyeos.

Apologies, upon further inspection, it seems this works on the "sending" side (a streaming motionEye) but still does not work on the receiving side (motioneEye "hub" writing to file with "H.264/V4L2M2M" setting and manually compiled rpi-ffmpeg).

**Comment by MichaIng** [comment]
And now, for RPi camera modules which expose the `/dev/video*` interface for libcamera rather than plain V4L2, do this: https://github.com/motioneye-project/motioneye/issues/2683#issuecomment-2207622479

**Comment by ToeKneeFan** [comment]
@MichaIng 
> And now, for RPi camera modules which expose the `/dev/video*` interface for libcamera rather than plain V4L2, do this: [#2683 (comment)](https://github.com/motioneye-project/motioneye/issues/2683#issuecomment-2207622479)

Unfortunately, that did not rescue functionality for me. Actually, my camera is a bit on the older side, so it would precede the libcamera-based devices.

Ah, after a bit of testing, I was able to get it to work! I noticed that ffmpeg with `-c:v h264_v4l2m2m` would work with sudo but fail with a regular user (whereas software encoding `h264` would work irrespective of sudo). This seems to be due to the fact that the HW-accelerated rendering device is owned by root:render.
Running `sudo adduser motion render` is sufficient to fix the permissions issue.

Presumably, the HW acceleration worked in the past because motionEye historically ran as root. Because motionEye is now run by the motion user, which does not have permission to use the HW-accelerated render device by default, HW acceleration fails.

I initially thought this was a motion or ffmpeg issue, but this does appear to be related to the setup of the motion user by motionEye. (The previously mentioned streaming devices were running ffmpeg as sudo, hence the discrepancy. Pardon the nonstandard setup. I was going to test on a clean installation if the above did not work.)

TL;DR motion user does not have permission to do hardware-accelerated encoding/decoding.

**Comment by MichaIng** [comment]
Oh, I would have assumed the `motion` user being part of the `render` group OOTB, as it is part of the `video` group. For some codecs with hardware acceleration this is indeed needed.

Good to know, so we could do that in the `motioneye_init` command.

> Actually, my camera is a bit on the older side, so it would precede the libcamera-based devices.

RPi camera modules von v3 on do not support the old API with `h264_omx` codec at all, but the other way round: older camera modules are exposed via libcamera API as well, isn't it? I have only a v4 here for testing, so if you have an older one, that is actually good to know and consider how it works:
- So it is a v1 or v2 RPi CSI camera module?
- How did you originally try to add it? As MMAL or V4L2 camera?
- And then you probably selected `/dev/video0`. Did the camera stream show up in the UI, and just recording did not work?
- Does it work now with `libcamerify` and/or without?

**Comment by ToeKneeFan** [comment]
That sounds great! Inclusion in `motioneye_init` makes sense (or even just mention in the installation steps if it would be easier).

- It is actually a third-party camera (LoveRPi Wide Angle Camera for Raspberry Pi, 170 FOV). It uses a OV5647 sensor, similar to the v1 CSI camera module. It seems to have been available starting in 2018, so approximately contemporaneous with the v1 module.
- I assume by "original" you mean for this recent installation, and not an older one (not my config on an old Raspbian OS with an old version of motioneye). Camera Type: "Local V4L2 Camera", Camera: "mmal service 16.1"
- Yes, without adding `motion` to `render`, it streams but does not record. Presumably because streaming relies on conversion to `mjpeg` rather than `h264_v4l2m2m`.
- Yes, it seems to work both with and without `libcamerify`.

**Comment by MichaIng** [comment]
Okay that is interesting. So this camera exposes a regular V4L2 interface, not libcamera, and probably was not MMAL before either? I.e. did it show up in `vcgencmd get_camera` in earlier RPi OS or DietPi (Bullseye and older)? Probably not every CSI camera on RPi behaves like that.

**Comment by ToeKneeFan** [comment]
@MichaIng 

Just to double check, I reflashed using an old image backup containing Raspbian Buster with motionEye v0.42.1.

`vcgencmd get_camera` yields
```
supported=1 detected=1, libcamera interfaces=0
```

It seems to have been set up as an MMAL camera
```
mmalcam_name vc.ril.camera
```

For the newer setup (going back to Diet Pi Trixie), it may be camera specific, though I wonder if it is software related. It specifically works through "mmal service." I believe this is just a sort of compatibility layer / virtual V4L2 device created by the `bcm2835-v4l2` kernel module to interact with MMAL devices (but don't quote me on that).

**Comment by MichaIng** [comment]
Okay, as this is added with `mmalcam_name`, it works with the legacy RPi camera driver. Yes `bcm2835-v4l2` did provide a V4L2-compatible interface, but with additional features like OMX.

I just recognized that motion 4.7.0 removed the `mmalcam_name` option, so we should remove the MMAL handling entirely as well.

**Comment by ToeKneeFan** [comment]
For sake of clarification, on the old Raspbian Buster with motionEye v0.42.1, it is set up as
```
mmalcam_name vc.ril.camera
```

On the new Diet Pi Trixie with motionEye dev, it is set up as
```
video_device /dev/video0
``` 

> I just recognized that motion 4.7.0 removed the `mmalcam_name` option, so we should remove the MMAL handling entirely as well.

Ah, that must be why the newer instance needs the V4L2 MMAL service instead of adding directly through `mmalcam_name`.

**Comment by MichaIng** [comment]
Yes `mmalcam_name` is used when adding it as MMAL camera, `video_device` when adding it as regualar V4L2 camera.

Which "V4L2 MMAL service" do you mean? Even if `motion` did still support it, the kernel does not, and the runtime libraries are not shipped anymore, hence it won't work in any case. `libcamerify` is currently the only way, unless the camera/driver does provide a regular V4L2 interface, instead of libcamera.

**Comment by ToeKneeFan** [comment]
> Which "V4L2 MMAL service" do you mean? Even if `motion` did still support it, the kernel does not, and the runtime libraries are not shipped anymore, hence it won't work in any case. `libcamerify` is currently the only way, unless the camera/driver does provide a regular V4L2 interface, instead of libcamera.

On the newer system, the camera is only accessible through Camera Type: "Local V4L2 Camera", Camera: "mmal service 16.1".

**Comment by ToeKneeFan** [comment]
@MichaIng 
I'm not sure, but, based on the git repo and some recent issues for the Raspberry Pi kernel, it seems many of those legacy components (namely, the MMAL service that creates a virtual V4L2-compatible interface) are still present and functional (but deprecated and officially unsupported) on recent _32-bit_ versions. Hence why it is working for me without `libcamerify`. (Pi Zero W runs the 32-bit kernel only.)

For example, [one user](https://github.com/raspberrypi/linux/issues/6753#issuecomment-2769013858) commented last year, "At some point it may get removed or just totally break." However, for now, it appears to still be working (but officially out of support).

**Comment by MichaIng** [comment]
> On the newer system, the camera is only accessible through Camera Type: "Local V4L2 Camera", Camera: "mmal service 16.1".

Ah, so as regular V4L2 camera, makes sense. These cameras always identify themselves as "mmal service x.y" in `v4l2-ctl --list-devices` output, but that does not necessarily mean much. I guess if `bcm2835-v4l2` is still loaded, and unloading it breaks the stream, then the old driver is still used under the hood, but no additional features that were used along with it, like OMX codec. The mainline V4L2 driver modules would be named `videobuf2-*`.

> For example, https://github.com/raspberrypi/linux/issues/6753#issuecomment-2769013858 commented last year, "At some point it may get removed or just totally break." However, for now, it appears to still be working (but officially out of support).

Even if you can still load the driver, the userspace library is missing. It was possible to download it manually from https://github.com/raspberrypi/firmware/blob/oldstable/hardfp/opt/vc/lib/libmmal.so, and IIRC `rpi-update` still used to install them as well, but above is oldstable Linux 6.1 branch, while in the stable/master/next it is gone: https://github.com/raspberrypi/firmware

Whichever way you turn it, MMAL is dead and we need to remove it entirely from motionEye code. Either a camera works as regular V4L2 camera, or it does so when using `libcamerify`. `motion` v5 will have native libcamera support, after which `libcamerify` is not needed anymore, and instead a new "libcamera" camera type aside of V4L2.
